activation/activation.go

```diff
@@ -897,20 +920,15 @@ func (b *Builder) getPositioningAtx(
 	publish types.EpochID,
 	previous *types.ActivationTx,
 ) (types.ATXID, error) {
-	id, err := b.searchPositioningAtx(ctx, nodeID, publish)
+	id, err := b.searchPositioningAtx(ctx, nodeID, publish, previous)
 	if err != nil {
 		return types.EmptyATXID, err
 	}
 
-	if previous != nil {
-		switch {
-		case id == b.conf.GoldenATXID:
-			id = previous.ID()
-		case id != b.conf.GoldenATXID:
-			if candidate, err := atxs.Get(b.db, id); err == nil {
-				if previous.TickHeight() >= candidate.TickHeight() {
-					id = previous.ID()
-				}
+	if previous != nil && id != previous.ID() {
+		if candidate, err := atxs.Get(b.db, id); err == nil {
```

Comment:
it is usually advised to keep happy path without nesting and handle errors in nested blocks. here, the happy path is nested, and errors aren't handled (swallowed). what happens if there's an error here? maybe for the sake of clarity it would be good to handle the error and explicitly set the variables so that the code becomes more readable (I understand that "this is how it was before the changeset" but maybe we can just improve it now)

Reply:
I do not know, why error is not returned here (and even isn't logged), but since getPositioningTAX returned some id and nil error, everything after is happy path, I think. I reason for nesting is avoiding repeating code with logging selected id and returning it. So I personally would keep it nested, but would log error in case if there is some

Reply:
The error here theoretically can't happen: we found an ATX for positioning that doesn't exist in our DB - ?

but @acud is right, the code can be written to avoid nesting, which makes it longer but easier to understand:

```go
	if previous != nil && id == previous.ID() {
		b.logger.Info("selected previous as positioning atx",
			log.ZShortStringer("id", id),
			log.ZShortStringer("smesherID", nodeID),
		)
		return id, nil
	}

	candidate, err := atxs.Get(b.db, id)
	if err != nil {
		return types.EmptyATXID, fmt.Errorf("fetching candidate positioning ATX %s from DB: %w", id.ShortString(), err)
	}

	if previous.TickHeight() >= candidate.TickHeight() {
		id = previous.ID()
	}

	b.logger.Info("selected positioning atx", log.ZShortStringer("id", id), log.ZShortStringer("smesherID", nodeID))
	return id, nil
``

Reply:
But then it will require to fix N tests, because we will need to add record to db. Do we want doing it in this PR? 
IMHO, it is refactoring stuff, which has to be in separate PR same as removal of unused veriables

Reply:
Which tests and how did they fail with the proposed change?

Reply:
- TestBuilder_PublishActivationTx_NoPrevATX: returns added error (can't fetch tx from ATX from db, becaose it doesn't exist)
- TestBuilder_PublishActivationTx_NoPrevATX_PublishFails_InitialPost_preserved - same
- TestWaitPositioningAtx - same
- TestGetPositioningAtx - same
- Test_Builder_RegenerateInitialPost - mocks failing because of unknown error
- Test_Builder_Multi_HappyPath - mocks failing because of unknown error (?)

Reply:
The failing tests actually indicate missed checks in my suggested code (that are currently done implicitly):

1. the code doesn't check if the selected positioning ATX is the golden ATX before fetching it from DB
2. if we found a `candidate` and `previous` is `nil` we get a `nil` pointer exception

To fix them we just need to adjust the code a bit:

```go
	id, err := b.searchPositioningAtx(ctx, nodeID, publish, previous)
	if err != nil {
		return types.EmptyATXID, err
	}

	if previous != nil && id == previous.ID() {
		b.logger.Info("selected previous as positioning atx",
			log.ZShortStringer("id", id),
			log.ZShortStringer("smesherID", nodeID),
		)
		return id, nil
	}

	if id == b.conf.GoldenATXID {
		b.logger.Info("selected golden atx as positioning atx", log.ZShortStringer("smesherID", nodeID))
		return id, nil
	}

	candidate, err := atxs.Get(b.db, id)
	if err != nil {
		return types.EmptyATXID, fmt.Errorf("get candidate pos ATX %s: %w", id.ShortString(), err)
	}

	if previous != nil && previous.TickHeight() >= candidate.TickHeight() {
		id = previous.ID()
		b.logger.Info("selected previous as pos atx",
			log.ZShortStringer("id", id),
			log.ZShortStringer("smesherID", nodeID),
		)
		return id, nil
	}

	b.logger.Info("selected positioning atx", log.ZShortStringer("id", id), log.ZShortStringer("smesherID", nodeID))
	return id, nil
```

with the additional checks all tests you listed pass for me 🙂

Reply:
Have changed code a little bit, according to this comment https://github.com/spacemeshos/go-spacemesh/pull/6053#discussion_r1650529338